libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+		/* TODO: DEBUG_CONTROL_C */
+		switch (code) {
+		case CREATE_PROCESS_DEBUG_EVENT:
+			//TODO: replace w32_h2t calls to w32_thread_list to get tid
+			/*eprintf ("(%d) created process (%d:%p)\n",
+				pid, w32_h2t (de.u.CreateProcessInfo.hProcess),
+				de.u.CreateProcessInfo.lpStartAddress);*/
+			w32_continue (dbg, pid, tid, -1);
+			next_event = 1;
+			ret = R_DEBUG_REASON_NEW_PID;
+			break;
+		case EXIT_PROCESS_DEBUG_EVENT:
+			//eprintf ("(%d) Process %d exited with exit code %d\n", (int)de.dwProcessId, (int)de.dwProcessId,
+			//	(int)de.u.ExitProcess.dwExitCode);
+			/*r_cons_printf ("(%d) Process %d exited with exit code %d\n", (int)de.dwProcessId, (int)de.dwProcessId,
+				(int)de.u.ExitProcess.dwExitCode);
+			r_cons_flush ();*/
+			//debug_load();
+			next_event = 0;
+			exited_already = pid;
+			ret = R_DEBUG_REASON_EXIT_PID;
+			break;
+		case CREATE_THREAD_DEBUG_EVENT:
+			//eprintf ("(%d) Created thread %d (start @ %p)\n", pid, tid, de.u.CreateThread.lpStartAddress);
+			r_debug_lstThreadAdd (pid, tid, de.u.CreateThread.hThread, de.u.CreateThread.lpThreadLocalBase, de.u.CreateThread.lpStartAddress, FALSE);
+			//w32_continue (dbg, pid, tid, -1);
+			ret = R_DEBUG_REASON_NEW_TID;
+			next_event = 0;
+			break;
+		case EXIT_THREAD_DEBUG_EVENT:
+			//eprintf ("(%d) Finished thread %d\n", pid, tid);
+			lstThreadPtr = (PTHREAD_ITEM)r_debug_findthread (pid, tid);
+			if (lstThreadPtr) {
+				lstThreadPtr->bFinished = TRUE;
+				lstThreadPtr->dwExitCode = de.u.ExitThread.dwExitCode;
+			} else {
+				r_debug_lstThreadAdd (pid, tid, de.u.CreateThread.hThread, de.u.CreateThread.lpThreadLocalBase, de.u.CreateThread.lpStartAddress, TRUE);
+			}
+			//w32_continue (dbg, pid, tid, -1);
+			next_event = 0;
+			ret = R_DEBUG_REASON_EXIT_TID;
+			break;
+		case LOAD_DLL_DEBUG_EVENT:
+			dllname = get_file_name_from_handle (de.u.LoadDll.hFile);
+			//eprintf ("(%d) Loading library at %p (%s)\n",pid, de.u.LoadDll.lpBaseOfDll, dllname ? dllname : "no name");
+			r_debug_lstLibAdd (pid,de.u.LoadDll.lpBaseOfDll, de.u.LoadDll.hFile, dllname);
+			if (dllname) {
+				free (dllname);
+			}
+			next_event = 0;
+			ret = R_DEBUG_REASON_NEW_LIB;
+			break;
+		case UNLOAD_DLL_DEBUG_EVENT:
+			//eprintf ("(%d) Unloading library at %p\n", pid, de.u.UnloadDll.lpBaseOfDll);
+			lstLibPtr = (PLIB_ITEM)r_debug_findlib (de.u.UnloadDll.lpBaseOfDll);
+			if (lstLibPtr != NULL) {
+				lstLibPtr->hFile = INVALID_HANDLE_VALUE;
+			} else {
+				r_debug_lstLibAdd (pid, de.u.UnloadDll.lpBaseOfDll, INVALID_HANDLE_VALUE, "not cached");
+				if (dllname)
+					free (dllname);
+			}
+			next_event = 0;
+			ret = R_DEBUG_REASON_EXIT_LIB;
+			break;
+		case OUTPUT_DEBUG_STRING_EVENT:
+			//eprintf ("(%d) Debug string\n", pid);
+			r_cons_printf ("(%d) Debug string\n", pid);
+			r_cons_flush ();
+
+			w32_continue (dbg, pid, tid, -1);
+			next_event = 1;
+			break;
+		case RIP_EVENT:
+			//eprintf ("(%d) RIP event\n", pid);
+			r_cons_printf ("(%d) RIP event\n", pid);
```

Comment:
we dont want debugger to flush anything, this will break r2pipe, we have an event API for those things. can you kepe the eprintf meanwhile?

Reply:
Yeah sure. I wasn't entirely sure why it was even made that way and I was a bit worried touching that in case that it breaks anything.